### PR TITLE
feat(doctor): config-time namespace-policy lint (#1888)

### PR DESCRIPTION
**Pull request by joshuaswarren**
## Context

Issue #1888 improvement 3 (config-time lint). Parts 1-2 dead-lettered ACL-rejected writes and surfaced the parked count in `doctor`; part 3 added the client startup preflight. This adds the **config-time lint** so the misconfiguration is caught before any write is attempted.

## Change

`remnic doctor` gains a **"Namespace policy"** check: when a configured write namespace matches no `namespacePolicies` entry and is neither the `defaultNamespace` nor the `sharedNamespace`, it is writable for nobody — every write to it is rejected by the namespace ACL and dead-lettered rather than stored. The check warns with remediation (add a policy entry, or use the default namespace).

Backed by a new principal-agnostic `isNamespacePolicyCovered(namespace, config)` helper exported from `@remnic/core` and unit-tested.

Part of #1888.

---

> [!NOTE]
> **Low Risk**
> Read-only diagnostic on config load in `doctor`; no changes to ACL enforcement or write paths at runtime.
> 
> **Overview**
> **`remnic doctor`** now runs a **Namespace policy** check when the raw config sets a client/plugin `namespace` write target. Invalid values (present but not a non-empty string) fail the check with remediation; valid values are checked against parsed namespace policy via new **`isNamespacePolicyCovered`** in `@remnic/core`.
> 
> That helper mirrors **`canWriteNamespace`** without a principal: with enforcement on, only the default namespace (when no matching policy) or a policy with at least one non-blank `writePrincipals` entry counts as writable—**`shared`** is not implicitly writable, and an explicit policy on `default` with no writers is not covered. Uncovered namespaces produce a **warning** with remediation so misconfigs that would dead-letter every write are caught before runtime.
> 
> Logic lives in **`doctor-namespace-lint.ts`** and is wired into standalone config parsing in **`cmdDoctor`**; unit tests cover precedence and edge cases.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 0554748c4db83cee6e6f6bf4a37c711b70083356. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

## Summary by CodeRabbit

* **New Features**
  * `remnic doctor` now adds a config-time **“Namespace policy”** diagnostic for the configured write namespace, including remediation when the `namespace` value is missing/invalid or when no enabled policy writers cover it.
  * Clarified policy coverage behavior: `default` can be covered when appropriate, while `shared` is not implicitly covered.
* **Bug Fixes**
  * Helps prevent unexpected namespace write rejection and dead-lettering by detecting coverage gaps before runtime.
* **Tests**
  * Added additional test coverage for namespace policy determination, including cases where enforcement is disabled.

### CODE REVIEW

File: packages/remnic-core/src/namespaces/principal.ts [hunk 1]
```diff
@@ -67,6 +67,23 @@ export function canWriteNamespace(principal: string | undefined, namespace: stri
...
+ * (no policy names it, and it is not the default), so every write to it is
+ * rejected by the ACL and dead-lettered. `remnic doctor` warns on that before
+ * a session silently loses memory.
+ */
+export function isNamespacePolicyCovered(namespace: string, config: PluginConfig): boolean {
+  return (
+    config.namespacePolicies.some((p) => p.name === namespace) ||
+    namespace === config.defaultNamespace ||
+    namespace === config.sharedNamespace
+  );
```
[the remark this answers is not in the document]
Comment (joshuaswarren) [reply packages/remnic-core/src/namespaces/principal.ts]: Fixed: isNamespacePolicyCovered now mirrors canWriteNamespace exactly — namespaces-off ⇒ all writable (no warning); otherwise covered only if the namespace is the defaultNamespace OR has a namespacePolicies entry that grants a writer. sharedNamespace is no longer treated as implicitly writable, and a policy with empty writePrincipals is not covered. Unit-tested for all cases (default/policy-with-writers covered; shared/no-writers/unlisted not; disabled ⇒ covered).

File: packages/remnic-cli/src/index.ts [hunk 2]
```diff
@@ -6573,6 +6581,26 @@ async function cmdDoctor(): Promise<void> {
...
+      ok: covered,
+      warn: !covered,
+      detail: covered
+        ? `configured namespace "${configuredNamespace}" is covered`
+        : `configured namespace "${configuredNamespace}" matches no namespacePolicies entry and is not the default/shared namespace`,
+      remediation: covered
+        ? undefined
+        : "Add a namespacePolicies entry (with your principal in writePrincipals), or set the namespace to the daemon's defaultNamespace — otherwise every write is rejected and dead-lettered.",
+    });
+  }
```
[the remark this answers is not in the document]
Comment (joshuaswarren) [reply packages/remnic-cli/src/index.ts]: Fixed: isNamespacePolicyCovered now mirrors canWriteNamespace exactly — namespaces-off ⇒ all writable (no warning); otherwise covered only if the namespace is the defaultNamespace OR has a namespacePolicies entry that grants a writer. sharedNamespace is no longer treated as implicitly writable, and a policy with empty writePrincipals is not covered. Unit-tested for all cases (default/policy-with-writers covered; shared/no-writers/unlisted not; disabled ⇒ covered).

File: packages/remnic-core/src/namespaces/principal.ts [hunk 3]
```diff
@@ -67,6 +67,23 @@ export function canWriteNamespace(principal: string | undefined, namespace: stri
...
+ * writable. A configured namespace that is NOT covered is writable for nobody
+ * (no policy names it, and it is not the default), so every write to it is
+ * rejected by the ACL and dead-lettered. `remnic doctor` warns on that before
+ * a session silently loses memory.
+ */
+export function isNamespacePolicyCovered(namespace: string, config: PluginConfig): boolean {
+  return (
+    config.namespacePolicies.some((p) => p.name === namespace) ||
+    namespace === config.defaultNamespace ||
+    namespace === config.sharedNamespace
```
[the remark this answers is not in the document]
Comment (joshuaswarren) [reply packages/remnic-core/src/namespaces/principal.ts]: Fixed: isNamespacePolicyCovered now mirrors canWriteNamespace exactly — namespaces-off ⇒ all writable (no warning); otherwise covered only if the namespace is the defaultNamespace OR has a namespacePolicies entry that grants a writer. sharedNamespace is no longer treated as implicitly writable, and a policy with empty writePrincipals is not covered. Unit-tested for all cases (default/policy-with-writers covered; shared/no-writers/unlisted not; disabled ⇒ covered).

File: packages/remnic-core/src/namespaces/principal.ts [hunk 4]
```diff
@@ -67,6 +67,23 @@ export function canWriteNamespace(principal: string | undefined, namespace: stri
...
+  );
+}
```
[the remark this answers is not in the document]
Comment (joshuaswarren) [reply packages/remnic-core/src/namespaces/principal.ts:93]: Fixed: isNamespacePolicyCovered now mirrors canWriteNamespace exactly — namespaces-off ⇒ all writable (no warning); otherwise covered only if the namespace is the defaultNamespace OR has a namespacePolicies entry that grants a writer. sharedNamespace is no longer treated as implicitly writable, and a policy with empty writePrincipals is not covered. Unit-tested for all cases (default/policy-with-writers covered; shared/no-writers/unlisted not; disabled ⇒ covered).

File: packages/remnic-cli/src/index.ts [hunk 5]
```diff
@@ -6521,11 +6522,18 @@ async function cmdDoctor(): Promise<void> {
...
+  let configuredNamespace: string | undefined;
   if (configExists) {
     try {
       const raw = JSON.parse(fs.readFileSync(configPath, "utf8"));
       const remnicCfg = resolveRemnicConfigRecord(raw);
       standaloneOpenaiApiKeyExplicitlyFalse = isOpenaiApiKeyDisabled(remnicCfg.openaiApiKey);
+      if (typeof remnicCfg.namespace === "string" && remnicCfg.namespace.trim().length > 0) {
+        configuredNamespace = remnicCfg.namespace.trim();
+      }
       standaloneConfig = parseConfig(remnicCfg);
```
[the remark this answers is not in the document]
Comment (joshuaswarren) [reply packages/remnic-cli/src/index.ts]: Fixed: a `namespace` key that is present but not a non-empty string is now surfaced as a FAILED "Namespace policy" check (with remediation) instead of being silently skipped.

File: packages/remnic-core/src/namespaces/principal.ts [hunk 6]
```diff
@@ -67,6 +67,25 @@ export function canWriteNamespace(principal: string | undefined, namespace: stri
...
+ * `canWriteNamespace`), and a policy entry with no `writePrincipals` grants
+ * nobody. A configured namespace that is NOT covered is writable for no one,
+ * so every write to it is rejected by the ACL and dead-lettered — `remnic
+ * doctor` warns on that before a session silently loses memory.
+ */
+export function isNamespacePolicyCovered(namespace: string, config: PluginConfig): boolean {
+  if (!resolveNamespaceCapabilities(config).namespaces) return true;
+  if (namespace === config.defaultNamespace) return true;
+  const policy = config.namespacePolicies.find((p) => p.name === namespace);
+  return policy !== undefined && policy.writePrincipals.length > 0;
```
[the remark this answers is not in the document]
Comment (joshuaswarren) [reply packages/remnic-core/src/namespaces/principal.ts]: Fixed: coverage now counts only non-blank writers (`policy.writePrincipals.some(w => w.trim().length > 0)`), so a policy with only blank entries like `[""]` — which `canWriteNamespace` can never match — is not treated as covered. Added a `blank-writer-ns` regression case asserting it is uncovered.

File: packages/remnic-cli/src/index.ts [hunk 7]
```diff
@@ -6573,6 +6589,33 @@ async function cmdDoctor(): Promise<void> {
...
+  // targets). An invalid `namespace` value is a failed check, not skipped.
+  if (configuredNamespaceInvalid) {
+    checks.push({
+      name: "Namespace policy",
+      ok: false,
+      detail: "config `namespace` is set but is not a non-empty string",
+      remediation: "Set `namespace` to a non-empty string (a namespacePolicies name or the default namespace), or remove it.",
+    });
+  } else if (standaloneConfig && configuredNamespace) {
+    const covered = isNamespacePolicyCovered(configuredNamespace, standaloneConfig);
```
[the remark this answers is not in the document]
Comment (joshuaswarren) [reply packages/remnic-cli/src/index.ts]: Intentionally NOT exempting sharedNamespace — this directly contradicts the higher-severity findings on the previous round (cursor High, coderabbit Major, codex P2), which I aligned with. `canWriteNamespace` does not special-case sharedNamespace: for a namespace with no policy it returns `namespace === config.defaultNamespace`, so `sharedNamespace` (which is not the default) is writable by NO ONE unless a policy grants it. Exempting it here would make `doctor` report a false OK while every write to `shared` is still ACL-rejected and dead-lettered — the precise bug the High-severity thread required fixing. If an operator wants `shared` writable, they add a namespacePolicies entry for it (e.g. writePrincipals:["*"]), and then the helper reports it covered. The helper deliberately mirrors canWriteNamespace exactly.

File: packages/remnic-core/src/namespaces/principal.ts [hunk 8]
```diff
@@ -67,6 +67,25 @@ export function canWriteNamespace(principal: string | undefined, namespace: stri
...
+ * `canWriteNamespace`), and a policy entry with no `writePrincipals` grants
+ * nobody. A configured namespace that is NOT covered is writable for no one,
+ * so every write to it is rejected by the ACL and dead-lettered — `remnic
+ * doctor` warns on that before a session silently loses memory.
+ */
+export function isNamespacePolicyCovered(namespace: string, config: PluginConfig): boolean {
+  if (!resolveNamespaceCapabilities(config).namespaces) return true;
+  if (namespace === config.defaultNamespace) return true;
+  const policy = config.namespacePolicies.find((p) => p.name === namespace);
+  return policy !== undefined && policy.writePrincipals.some((w) => w.trim().length > 0);
```
[the remark this answers is not in the document]
Comment (joshuaswarren) [reply packages/remnic-core/src/namespaces/principal.ts]: Fixed — real bug. The helper now mirrors `canWriteNamespace` precedence exactly: it looks up an explicit `namespacePolicies` entry first and only falls back to `namespace === defaultNamespace` when no policy matches. So a policy naming the default namespace with empty/blank-only `writePrincipals` is now correctly reported as NOT covered (writable by no one), instead of being short-circuited to covered. Added a regression test (`an explicit default-namespace policy overrides the default fallback`) covering empty, blank-only, and granted default-namespace policies.

File: packages/remnic-cli/src/doctor-namespace-lint.ts [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+  }
+  return { invalid: true };
+}
+
+/**
+ * Build the config-time namespace-policy lint row (issue #1888 improvement 3). A
+ * configured write namespace that is writable by no one (no default, no policy
+ * granting a writer) has every write rejected by the ACL and dead-lettered; catch
+ * it at config time before a session silently loses memory. Returns undefined when
+ * there is nothing to lint (no configured namespace and the value was not invalid).
```
[the remark this answers is not in the document]
Comment (joshuaswarren) [reply packages/remnic-cli/src/doctor-namespace-lint.ts:40]: Fixed — the message and remediation were stale after the precedence fix. The not-covered detail no longer claims "is not the default namespace" (which is false for a default namespace shadowed by an empty policy); it now reads "its namespacePolicies entry grants no writer, or it has no entry and is not the default namespace", and the remediation points to adding a non-blank writePrincipals entry for that specific namespace (or choosing a writable one) rather than "set it to defaultNamespace" — which would not help when the default itself carries a writerless policy.
